### PR TITLE
generate linnarsson zarr from hdf5 directly

### PR DESCRIPTION
**Pull request by manzt**
We don't need to convert to OME to then convert back to zarr... 

I added scripts to convert directly to zarr and then tile the chunked full resolution array into lower resolutions. Since we haven't determined our standard yet for pyramids (see https://github.com/hubmapconsortium/vitessce-image-viewer/issues/91), these scripts generate tiles in the format current supported for zarr in Viv: 

```bash
.
└── linnarsson.images.zarr 
    └── pyramid 
        ├── 00 # generated by img_hdf5_reader.py
        │   ├── .zarray
        │   ├── 0.0.0
        │   └── ..etc
        ├── 01 # generated by tile_zarr_base.py
        │   ├── .zarray
        │   ├── 0.0.0
        │   └── ..etc
        └── 02 # generated by tile_zarr_base.py
            ├── .zarray
            ├── 0.0.0
            └── ..etc
```

The `linnarsson.images.json` should be _very_ similar to the config in viv now, where keys here are "channels" and have the same url since it's one zarr array:

```
{
  "polyT": {
    "sample": 1,
    "tileSource": "https://s3.amazonaws.com/vitessce-data/0.0.20/master_release/linnarsson/linnarsson.images.zarr/pyramid/"
  },
  "nuclei": {
    "sample": 1,
    "tileSource": "https://s3.amazonaws.com/vitessce-data/0.0.20/master_release/linnarsson/linnarsson.images.zarr/pyramid/"
  }
}
```

The only thing missing in Viv logic is `minZoom`. I guess I could add that here to the `linnarsson.images.json`. In the future we shouldn't be exporting a `linnarsson.images.json` and instead should have this contained in the `.zattrs` of the base zarr. But this is what viv supports now and we want to update our demo. It would be good to iron out the zarr schema sooner rather than later..

### CODE REVIEW

**Comment by manzt** [comment]
We will need to change the s3 urls obviously, but all the converters are in place.

**Comment by mccalluc** [comment]
> We don't need to convert to OME to then convert back to zarr...

What I remember from last summer was that we wanted to think of OME-TIFF as our central format: We think we'll have a lot of formats coming in, and we may want to generate a range of output formats as well, and we don't want m*n scripts.

This assumptions might no longer be true, but I'd like to be more clear about the roadmap. Perhaps loop in Nils?

Sorry not to raise this concern earlier.

**Comment by manzt** [comment]
I am working on making converters for many types of images, but specifically OME-TIFF with Heath in https://github.com/hubmapconsortium/img2zarr. I figured for the purposes of vitessce, and @ngehlenborg demo on wednesday, this should suffice because its the format the data are provided in.

File: fake-files/output-expected/linnarsson/linnarsson.images.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
   "polyT": {
     "sample": 1,
-    "tileSource": "https://s3.amazonaws.com/vitessce-data/0.0.20/master_release/linnarsson/linnarsson.tiles/linnarsson.images.polyT/polyT.dzi"
+    "tileSource": "https://s3.amazonaws.com/vitessce-data/0.0.20/master_release/linnarsson/linnarsson.images.zarr/pyramid/"
```
Comment (mccalluc) [inline fake-files/output-expected/linnarsson/linnarsson.images.json]: Can you update the directory we're pushing to?

File: python/img_hdf5_reader.py [hunk 2]
```diff
@@ -58,12 +66,12 @@ def scale_sample(self, channel, sample, max_allowed, clip):
...
 
         '''
-        sampled = np.clip(self.sample_image(channel, sample), 0, clip)
+        sampled = self.sample_image(channel, sample, use_dask).clip(0, clip)
         # 255 displays as black... color table issue?
         return sampled / clip * (max_allowed - 1)
 
     def to_png(self, channel, sample, png_path, clip):
-        MAX_ALLOWED = 256
+        MAX_ALLOWED = 65536
```
Comment (mccalluc) [inline python/img_hdf5_reader.py]: Are we still making PNGs? Maybe get rid of this function? I don't know if 16-bit PNGs are a thing.

File: python/img_hdf5_reader.py [hunk 3]
```diff
@@ -187,6 +195,74 @@ def to_ometiff(self, channel_clips, sample, json_file):
...
+        # zarr default BLOSC not supported in browser
+        COMPRESSOR = Zlib(level=1)
+        MAX_ALLOWED = 65536
+        TILE_SIZE = 512
+
+        channels = [channel for (channel, clip) in channel_clips]
+        data_shape, data_dtype = self._get_shape_and_dtype(channels)
+
+        # Create mutable store
+        zarr_path = f"{json_file.name.replace('.json', '')}.zarr"
```
Comment (mccalluc) [inline python/img_hdf5_reader.py]:
```suggestion
        zarr_path = json_file.name.replace('.json', '.zarr')
```

File: python/img_hdf5_reader.py [hunk 4]
```diff
@@ -187,6 +195,74 @@ def to_ometiff(self, channel_clips, sample, json_file):
...
+        da.array(images).rechunk(chunks).to_zarr(
+            zarr_path, component=f"{pyramid_group}/00", compressor=COMPRESSOR
+        )
+        json.dump(channel_data, json_file, indent=2)
+
+    def _get_shape_and_dtype(self, channels):
+        shapes, dtypes = zip(
+            *[(self.data[c].shape, self.data[c].dtype) for c in channels]
+        )
+        # Compare dimesion sizes to make sure all same size
```
Comment (mccalluc) [inline python/img_hdf5_reader.py]:
```suggestion
        # Compare dimension sizes to make sure all same size
```

File: python/img_hdf5_reader.py [hunk 5]
```diff
@@ -187,6 +195,74 @@ def to_ometiff(self, channel_clips, sample, json_file):
...
+            if not all_same:
+                raise ValueError(
+                    f"Data stored in {channels} are not the same shape."
+                )
+            shape.append(first_el)
+        # Ensure same dtype.
+        first_dtype = dtypes[0]
+        all_same = all(dtype == first_dtype for dtype in dtypes)
+        if not all_same:
+            raise ValueError(f"Dtypes are not the same for {channels}.")
```
Comment (mccalluc) [inline python/img_hdf5_reader.py]: This is clean, but would it be better to iterate over the dtypes, so if there was a problem, we could say exactly where it was? Though if that's unlikely, not worth the change.

File: python/tile_zarr_base.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+        if img.shape[2] < TILE_SIZE:
+            img = da.pad(
+                img,
+                ((0, 0), (0, 0), (0, TILE_SIZE - img.shape[2])),
+                "constant",
+            )
+
+        # Define pyramid level path
+        out_path = str(pyramid_path / str(i).zfill(2))
```
Comment (mccalluc) [inline python/tile_zarr_base.py:55]: I think I'd prefer if our `_path` variables were Paths.

File: python/img_hdf5_reader.py (hunk 3 above)
Comment (mccalluc) [reply python/img_hdf5_reader.py]: Or better, use pathlib,
```
zarr_path = json_file.with_suffix('.zarr')
```

File: scripts/process_linnarsson.sh [hunk 7]
```diff
@@ -104,23 +104,18 @@ process_linnarson_images() {
...
-
-        CMD='docker run --rm
-            -e "CHANNEL_PAGE_PAIRS=polyT:0 nuclei:1"
-            -e "PREFIX=linnarsson"
-            --mount "type=bind,src='$OUTPUT'/linnarsson.images.ome.tif,destination=/input.ome.tif"
-            --mount "type=bind,src='$TILES_PATH',destination=/output_dir"
-            --name tiler gehlenborglab/ome-tiff-tiler:v0.0.3'
+         URL_PREFIX="https://s3.amazonaws.com/$S3_TARGET/linnarsson/$TILES_BASE"
+
+    CMD="$BASE/python/tile_zarr_base.py
```
Comment (mccalluc) [inline scripts/process_linnarsson.sh]: Indent

File: test.sh [hunk 8]
```diff
@@ -40,11 +40,6 @@ CI=true ./process.sh
 # CI to get it to run as test locally;
 # Redundant on Travis, but doesn't hurt anything.
 
-perl -pne \
-  's/<ome:AcquisitionDate>.{26}/<ome:AcquisitionDate>0000-00-00T00:00:00.000000/;
-	s/tifffile.py.{20}/tifffile.py.0000:00:00 00:00:00/' -i fake-files/output/linnarsson/linnarsson.images.ome.tif
-
-
 # If you have problems with the TIFF, try this:
```
Comment (mccalluc) [inline test.sh]: And you can delete the debugging help below, too!

**Review by mccalluc — COMMENTED** [review]
Looks good. I think we can use pathlib more, but not a big deal. I'm not sure it needs to happen with every branch, but for this, can you update the s3 target? Thanks! Please merge once the couple items are addressed.

File: python/tile_zarr_base.py (hunk 6 above)
Comment (manzt) [reply python/tile_zarr_base.py:55]: Unfortunately zarr and dask.array don't accept `Path` objects. I use pathlib to ensure that the paths can be cross platform, and I believe the `Path.__str__`takes care of creating the correct path for the file system.

File: python/img_hdf5_reader.py (hunk 3 above)
Comment (manzt) [reply python/img_hdf5_reader.py]: `json_file` is not a Path object. 

```AttributeError: '_io.TextIOWrapper' object has no attribute 'with_suffix'```

File: fake-files/output-expected/linnarsson/linnarsson.images.json (hunk 1 above)
Comment (manzt) [reply fake-files/output-expected/linnarsson/linnarsson.images.json]: I would do this, but I'm not sure what that should be in google cloud! 

I pushed the data to `https://vitessce-data.storage.googleapis.com/linnarsson.images.zarr/pyramid` to test with the current version of viv, but I think we should version this end-point.

File: python/img_hdf5_reader.py (hunk 5 above)
Comment (manzt) [reply python/img_hdf5_reader.py]: I think this is unlikely

Comment (manzt) [reply python/img_hdf5_reader.py]: I changed this to print both the `channels` and `dtypes` so it's easier to debug.

File: python/tile_zarr_base.py (hunk 6 above)
Comment (manzt) [reply python/tile_zarr_base.py:55]: @mccalluc To clarify, `pyramid_path` is a Path but `out_path` must be a string for zarr and dask array. I could use `str(out_path)` in the `da.to_zarr` and `da.from_zarr` if that is preferable.

**Comment by ilan-gold** [comment]
We should talk tomorrow, but the schema in this PR and the Vanderbilt one should match.  I can show you what it should look like and you can also give me feedback since one is Zarr and one is not.

File: fake-files/output-expected/linnarsson/linnarsson.images.json (hunk 1 above)
Comment (ilan-gold) [reply fake-files/output-expected/linnarsson/linnarsson.images.json]: Add other necessary metaddata (zoom level)

Comment (ilan-gold) [reply fake-files/output-expected/linnarsson/linnarsson.images.json]: For URL: https://github.com/hubmapconsortium/vitessce-data/pull/73/files

File: fake-files/output-expected/linnarsson/linnarsson.images.zarr/pyramid/00/.zarray [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    "chunks": [
+        2,
+        50,
+        50
+    ],
+    "compressor": {
+        "id": "zlib",
+        "level": 1
+    },
+    "dtype": "<i4",
```
Comment (ilan-gold) [inline fake-files/output-expected/linnarsson/linnarsson.images.zarr/pyramid/00/.zarray:11]: Can you figure out what is going on here?  This should be `<u4` as far as I understand

Comment (manzt) [reply fake-files/output-expected/linnarsson/linnarsson.images.zarr/pyramid/00/.zarray:11]: This is a result of the data generated for the fixture (which is `<i4`). See here: 

https://github.com/hubmapconsortium/vitessce-data/blob/947bb6160dccc4afb967d05c3558a5bd7f641bff/python/create_hdf5_fixtures.py#L3-L10

Comment (manzt) [reply fake-files/output-expected/linnarsson/linnarsson.images.zarr/pyramid/00/.zarray:11]: The actual linnarsson data is u16 and that is respected.

Comment (ilan-gold) [reply fake-files/output-expected/linnarsson/linnarsson.images.zarr/pyramid/00/.zarray:11]: Ah, ok.  Sounds good.

**Review by ilan-gold — APPROVED** [review]

File: scripts/process_linnarsson.sh [hunk 10]
```diff
@@ -95,33 +95,30 @@ process_linnarson_images() {
...
+        ZARR_OUT='linnarsson.images.zarr'
+        ZARR_URL="https://vitessce-data.storage.googleapis.com/$RELEASE/linnarsson/$ZARR_OUT"
+
         CMD="$BASE/python/img_hdf5_reader.py
             --hdf5 $HDF5_IN
-            --channel_clip_pairs polyT:200 nuclei:20
             --sample 1
-            --json_file $JSON_OUT"
+            --channels polyT,nuclei
+            --json_file $JSON_OUT
```
Comment (ilan-gold) [inline scripts/process_linnarsson.sh:107]: Can you change `$JSON_OUT` to be `linnarsson.raster.json` instead of `linnarsson.images.json`?

Comment (ilan-gold) [reply scripts/process_linnarsson.sh:107]: Sorry I forgot about this, but I would rather not maintain two schemas.  I can change the file name on s3.

Comment (manzt) [reply scripts/process_linnarsson.sh:107]: I asked about this earlier ... sure

Comment (ilan-gold) [reply scripts/process_linnarsson.sh:107]: Yes, sorry about that.

Comment (manzt) [reply scripts/process_linnarsson.sh:107]: it wont be nested in the zarr storage. is that alright?

Comment (manzt) [reply scripts/process_linnarsson.sh:107]: `vitessce-data/0.0.23/master_release/linnarsson/linnarsson.raster.json`

File: fake-files/output-expected/linnarsson/linnarsson.raster.json [hunk 11]
```diff
@@ -1,7 +1,7 @@
 {
   "polyT": {
     "sample": 1,
-    "tileSource": "https://vitessce-data.storage.googleapis.com/0.0.20/master_release/linnarsson/linnarsson.images.zarr/pyramid/",
+    "tileSource": "https://vitessce-data.storage.googleapis.com/0.0.20/master_release/linnarsson//Users/trevormanz/GitHub/hubmap/vitessce-data/fake-files/output/linnarsson/linnarsson.images.zarr/pyramid/",
```
Comment (mccalluc) [inline fake-files/output-expected/linnarsson/linnarsson.raster.json]: This path isn't right.

File: process.sh [hunk 12]
```diff
@@ -35,9 +35,8 @@ main() {
     then
         echo 'CI: Skip push to AWS and GCS'
     else
-        TILES_BASE='vanderbilt.images'
-        aws s3 cp --exclude "$OUTPUT/*.ome.tif*" --exclude "$OUTPUT/*.zarr" --recursive "$OUTPUT" s3://"$S3_TARGET"
-        gsutil cp "$OUTPUT/vanderbilt/$TILES_BASE/*.ome.tif*" gs://vitessce-data/$TILES_BASE
+        aws s3 cp --exclude "$OUTPUT/*.ome.tif*" --exclude "$OUTPUT/*.zarr/*" --recursive "$OUTPUT" s3://"$S3_TARGET"
+        gsutil -m cp -r "$OUTPUT/vanderbilt/vanderbilt.images" gs://"$S3_TARGET/vanderbilt/vanderbilt.images"
```
Comment (mccalluc) [inline process.sh]:
```suggestion
        gsutil -m cp -r "$OUTPUT/vanderbilt/vanderbilt.images" "gs://$S3_TARGET/vanderbilt/vanderbilt.images"
```
No change in behavior, but quote the whole string for consistency.

(Perhaps rename S3_TARGET? Not a big deal.)

File: scripts/process_linnarsson.sh [hunk 13]
```diff
@@ -96,21 +97,21 @@ process_linnarson_images() {
...
+            --zarr_file $ZARR_OUT
+            --tiles_url $TILES_URL"
         echo "Running: $CMD"
         $CMD
     fi
 
-    TILES_BASE="$ZARR_OUT/pyramid"
-    TILES_PATH="$OUTPUT/$TILES_BASE"
+    TILES_PATH="$ZARR_OUT/pyramid"
+    echo "$TILES_PATH/01"
```
Comment (mccalluc) [inline scripts/process_linnarsson.sh]: Log statements are fine, but add some explanation of what this string is.

**Review by mccalluc — CHANGES_REQUESTED** [review]
I think a bad path got in here by mistake.

File: fake-files/output-expected/linnarsson/linnarsson.raster.json (hunk 11 above)
Comment (manzt) [reply fake-files/output-expected/linnarsson/linnarsson.raster.json]: fixed.

Comment (manzt) [reply fake-files/output-expected/linnarsson/linnarsson.raster.json]: Fixed.

File: process.sh (hunk 12 above)
Comment (manzt) [reply process.sh]: Done. I'll open an issue to rename S3_TARGET.

File: scripts/process_linnarsson.sh (hunk 13 above)
Comment (manzt) [reply scripts/process_linnarsson.sh]: This is vestigial. Removed.

**Comment by manzt** [comment]
@mccalluc I updated `s3_target.txt` --> `cloud_target.txt` and all instances of `$S3_TARGET` --> `$CLOUD_TARGET`

**Pull request merged**